Merge coverage results from multiple runs using unique command names

BaxterStockman opened the pull request:
This is a work-in-progress branch that addresses the result-merging issue raised in #28.

@infertux -- I'm not sure if I'm overdoing it with the introduction of feature tests.  I figure it's not a big deal from the point of view of introducing new dependencies (since SimpleCov already depends on Cucumber and Aruba), and the plain-language approach felt right for describing the kinds of behaviours introduced by these changes.  But it should be fairly straightforward to rework the specs in plain old RSpec, too.

The changesets are in a rough-ish state at the moment; I'll clean everything up and rebase if and when I get the go-ahead.  Thanks in advance!

infertux commented:
> @infertux -- I'm not sure if I'm overdoing it with the introduction of feature tests. I figure it's not a big deal from the point of view of introducing new dependencies (since SimpleCov already depends on Cucumber and Aruba)

@BaxterStockman That's totally fine by me. Those are just development dependencies. I try to limit *runtime* dependencies but this doesn't apply to dev dependencies ;)

> and the plain-language approach felt right for describing the kinds of behaviours introduced by these changes

Agreed.

> The changesets are in a rough-ish state at the moment; I'll clean everything up and rebase if and when I get the go-ahead. Thanks in advance!

No need to ask. Thank *you*! I don't really give this project the time it deserves lately. Truth be told, I think I'm getting a bit tired of writing Ruby code so I really appreciate your help here 😃 

Please ping me if you need a second pair of eyes to review ;)

infertux commented on bin/bashcov:
```diff
@@ -14,8 +14,15 @@ coverage = runner.result
-SimpleCov.at_exit { SimpleCov::Result.new(coverage).format! }
+
+result = SimpleCov::Result.new(coverage)
+if SimpleCov.use_merging
```
Maybe there should be a quick mention about this in the README?

BaxterStockman replied:
Agreed, and added :smiley:

infertux commented on lib/bashcov.rb:
```diff
@@ -68,6 +68,13 @@ def fullname
       ].join(" ")
     end
 
+    # @return [String] The value to
```
Missing the end of the sentence?

BaxterStockman commented:
Thanks for the feedback, @infertux!  I have addressed your diff comments and also added a changelog entry.  I'd sure appreciate a review of the README additions to make sure they adequately explain the purpose of the `--command-name` option and `BASHCOV_COMMAND_NAME` environment variable.  Other than that, I think this is probably ready for merging :)

infertux commented:
The README looks great, thanks @BaxterStockman.

However I just noticed something weird where the `SimpleCov.at_exit` block isn't being called for some reason:

`cd spec/test_app; rm -rf coverage; ../../bin/bashcov -- ./test_suite.sh`

The output doesn't show the coverage report was generated and indeed the `coverage/` directory is empty. Any ideas what's wrong? This happens both on Ruby 2.4 and 2.5.

BaxterStockman commented:
@infertux -- ouch!  I'm able to reproduce the issue, and it looks like the root cause is that 

1. The `Kernel.at_exit` block in `simplecov/defaults` short-circuits if `SimpleCov.pid != Process.pid`, and
2. `SimpleCov.pid` is only set in one place, `SimpleCov.start`, which Bashcov does not call.

Here's a `pry` session that displays the problem:

```ruby
bashcov: warning: /bashcov/spec/test_app/tmp.sh was executed but has been deleted since then - it won't be reported in coverage.

From: /home/travis/.rvm/gems/ruby-2.4.2/gems/simplecov-0.15.1/lib/simplecov/defaults.rb @ line 55 :

    50: # Gotta stash this a-s-a-p, see the CommandGuesser class and i.e. #110 for further info
    51: SimpleCov::CommandGuesser.original_run_command = "#{$PROGRAM_NAME} #{ARGV.join(' ')}"
    52:
    53: at_exit do # rubocop:disable Metrics/BlockLength
    54: require 'pry'
 => 55: binding.pry
    56:   # If we are in a different process than called start, don't interfere.
    57:   next if SimpleCov.pid != Process.pid
    58:
    59:   @exit_status = if $! # was an exception thrown?
    60:                    # if it was a SystemExit, use the accompanying status

[1] pry(main)> show-method SimpleCov.pid

From: /home/travis/.rvm/gems/ruby-2.4.2/gems/simplecov-0.15.1/lib/simplecov.rb @ line 22:
Owner: #<Class:SimpleCov>
Visibility: public
Number of lines: 1

attr_accessor :pid
[2] pry(main)> show-method SimpleCov.start

From: /home/travis/.rvm/gems/ruby-2.4.2/gems/simplecov-0.15.1/lib/simplecov.rb @ line 41:
Owner: #<Class:SimpleCov>
Visibility: public
Number of lines: 14

def start(profile = nil, &block)
  if SimpleCov.usable?
    load_profile(profile) if profile
    configure(&block) if block_given?
    @result = nil
    self.running = true
    self.pid = Process.pid
    Coverage.start
  else
    warn "WARNING: SimpleCov is activated, but you're not running Ruby 1.9+ - no coverage analysis will happen"
    warn "Starting with SimpleCov 1.0.0, even no-op compatibility with Ruby <= 1.8 will be entirely dropped."
    false
  end
end
[3] pry(main)> SimpleCov.pid
=> nil                      
```
I think that the solution may be to just set `SimpleCov.pid = Process.pid` in `bashcov`.  We also need to set `SimpleCov.running = true` to avoid hitting the short-circuiting condition in `SimpleCov.at_exit`

```ruby
=> nil
[4] pry(main)> show-method SimpleCov.at_exit
                                                                                                                                                                                                                                                                                                                              
From: /home/travis/.rvm/gems/ruby-2.4.2/gems/simplecov-0.15.1/lib/simplecov/configuration.rb @ line 177:
Owner: SimpleCov::Configuration
Visibility: public
Number of lines: 5
                                                                                                                                                                                                                                                                                                                              
def at_exit(&block)
  return proc {} unless running || block_given?
  @at_exit = block if block_given?
  @at_exit ||= proc { SimpleCov.result.format! }
end
```

Alternatively, we could do:

```ruby
SimpleCov.start do
  at_exit { result.format! }
end
```

Although there's more stuff going on in `SimpleCov.start` (like calling `Coverage.start`) that we don't necessarily want to trigger.

infertux commented:
Great stuff @BaxterStockman.

> SimpleCov.pid is only set in one place, SimpleCov.start, which Bashcov does not call.

Hmm not sure why we were not calling `SimpleCov.start` in the first place. I thought we were.

> I think that the solution may be to just set SimpleCov.pid = Process.pid in bashcov. We also need to set SimpleCov.running = true to avoid hitting the short-circuiting condition in SimpleCov.at_exit

This feels brittle to me as it will break Bashcov if SimpleCov changes the logic around there.

> Although there's more stuff going on in SimpleCov.start (like calling Coverage.start) that we don't necessarily want to trigger.

I don't see anything harmful in calling `SimpleCov.start`, were you referring to something in particular?

I tried that in e5d2e38 and it's working great for me.

infertux commented:
I noticed another issue. The features ran by Cucumber do not contribute to the overall Bashcov coverage. For example, this block is marked as missed after running `bundle exec rake` even though it's actually covered:

```ruby
opts.on("--command-name NAME", "Value to use as SimpleCov.command_name") do |c|
    options.command_name = c                                           
end
```

I think it's because of the `next if SimpleCov.pid != Process.pid` line you mentioned above. I tried [use this](https://relishapp.com/philoserf/aruba/docs/command/run-commands-in-ruby-process) but it gets really messy/hacky.

Maybe you can think of a clean way to do it? Otherwise I think it's fine to merge this PR as is.

infertux commented on features/step_definitions/bashcov_steps.rb:
```diff
@@ -0,0 +1,96 @@
+  end
+
+  def simplecov_results_json
+    run_simple(<<-'COMMAND', fail_on_error: false)
```
BTW do we actually need the `fail_on_error: false` here?

BaxterStockman replied:
Hm.  I remember that there was a good reason for passing that option at some point, but I've removed it and the features work correctly so I guess that reason no longer obtains.

BaxterStockman commented:
> I don't see anything harmful in calling SimpleCov.start, were you referring to something in particular?

One concern I had is that a user might do something like 

```ruby
# .simplecov
SimpleCov.configure do
  filters.clear
end
```

and end up with code from `bin/bashcov` in their coverage results.  However, I tested emptying out SimpleCov's filters, and didn't end up with extraneous code.  I'm guessing it's because Bashcov overrides the default `SimpleCov.at_exit` proc, so `SimpleCov.result` (which is what would include in-process Ruby coverage) never gets generated.

> The features ran by Cucumber do not contribute to the overall Bashcov coverage[...] I think it's because of the `next if SimpleCov.pid != Process.pid` line you mentioned above.

That sounds right on the money to me.  No bright ideas for a non-hacky in-process executor occur to me right away; however, the fact that [the link you posted](https://relishapp.com/philoserf/aruba/docs/command/run-commands-in-ruby-process) mentions SImpleCov explicitly makes me think that this is a common use-case and there might be a drop-in solution in the wild already.  I'll take a look around and, if I can't find anything promising, I'll ruminate on a roll-your-own executor.

infertux commented:
> [...] makes me think that this is a common use-case and there might be a drop-in solution in the wild already. I'll take a look around and, if I can't find anything promising, I'll ruminate on a roll-your-own executor.

I found [this](https://github.com/cucumber/aruba/issues/254) so presumably there is no drop-in solution.

Don't worry about it, it's not a big deal. I don't think it's worth building a hacky in-process runner just for that.

Merging this now. Thank you for the hard work. You're awesome! 😃

BaxterStockman commented:
> I don't think it's worth building a hacky in-process runner just for that.

Fair enough! The interface that Aruba provides for developing an in-process runner certainly leaves a more than a bit to be desired.  Also, following [the approach recommended by SimpleCov's author](https://github.com/colszowka/simplecov/issues/234#issuecomment-21293280) would amount to writing a program that does `require "simplecov"`, generates a unique command name, and merges the coverage results into existing results, all in order to... test a program that does `require "simplecov"`, generates a unique command name, and merges the coverage results into existing results (i.e., `bashcov` itself).

> Merging this now.

Excellent! TYVM :grin:
